src/crate_graph.rs

```diff
@@ -405,6 +454,67 @@ impl DependencyGraph {
         }
         visited
     }
+
+    pub fn detect_cycles(&self) -> Option<Vec<PackageId>> {
+        let mut visited = HashSet::new();
+        let mut recursion_stack = HashSet::new();
```

Comment:
Might help to clarify with a comment that `visited` is used as an optimization to prevent duplicating traversals, while `recursion_stack` is functionally necessary for actually detecting a cycle within one traversal.